<launch>
  <!-- Arguments -->
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  <arg name="cmd_vel_topic" default="/cmd_vel" />
  <arg name="odom_topic" default="odom" />
  <arg name="move_forward_only" default="false"/>
  <arg name="ns"/>
  <arg name="followers"/>

  <!-- move_base -->

  <group ns="$(arg ns)">
    <param name="tf_prefix" value="$(arg ns)" />
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" />
    
    <rosparam file="$(find turtlebot3_navigation)/param/costmap_common_params_$(arg model).yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find turtlebot3_navigation)/param/costmap_common_params_$(arg model).yaml" command="load" ns="local_costmap" />
    
    <rosparam file="$(find turtlebot3_navigation)/param/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find turtlebot3_navigation)/param/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find turtlebot3_navigation)/param/move_base_params.yaml" command="load" />
    
    <rosparam file="$(find turtlebot3_navigation)/param/dwa_local_planner_params_$(arg model).yaml" command="load" />
    
    <remap from="cmd_vel" to="$(arg cmd_vel_topic)"/>
    <remap from="odom" to="$(arg odom_topic)"/>
    <param name="DWAPlannerROS/min_vel_x" value="0.0" if="$(arg move_forward_only)" />


    <param name="global_costmap/global_frame" value="map_merge/map"/>
    <param name="global_costmap/static_map" value="true"/>
    <param name="global_costmap/rolling_window" value="true"/>
    <param name="global_costmap/robot_base_frame" value="$(arg ns)/base_footprint"/>
    <param name="local_costmap/robot_base_frame" value="$(arg ns)/base_footprint"/>
    <param name="local_costmap/global_frame" value="map_merge/map"/>
    <param name="local_costmap/width" value="1"/>
    <param name="local_costmap/height" value="1"/>
    <param name="local_costmap/track_unknown_space" value="true"/>

  </node>

  </group>

  <!-- recursive launch for followers -->
<arg name="new_followers" value="$(eval arg('followers') - 1)" />
<include file="$(find swarmbots)/launch/move_base.launch" if="$(eval arg('new_followers') >= 0)">
    <arg name="model" value="$(arg model)" />
    <arg name="move_forward_only" value="$(arg move_forward_only)"/>
    
    <arg name="followers" value="$(arg new_followers)"/>
    <arg name="ns" value="robot$(arg new_followers)"/>
</include>



</launch>
